facilitator/src/transport/s3.rs

```diff
@@ -764,3 +766,153 @@ mod tests {
+///
+/// The following sources are checked in order for credentials when calling `credentials`:
+///
+/// 1. Environment variables: `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`
+/// 2. `credential_process` command in the AWS config file, usually located at `~/.aws/config`.
+/// 3. AWS credentials file. Usually located at `~/.aws/credentials`.
+/// 4. IAM instance profile. Will only work if running on an EC2 instance with an instance profile/role.
+///
+/// If the sources are exhausted without finding credentials, an error is returned.
+///
+/// The provider has a default timeout of 30 seconds. While it should work well for most setups,
+/// you can change the timeout using the `set_timeout` method.
+///
+/// # Example
+///
+/// ```rust
+/// use std::time::Duration;
+///
+/// use rusoto_credential::ChainProvider;
+///
+/// let mut provider = ChainProvider::new();
+/// // you can overwrite the default timeout like this:
+/// provider.set_timeout(Duration::from_secs(60));
+/// ```
+///
+/// # Warning
+///
+/// This provider allows the [`credential_process`][credential_process] option in the AWS config
+/// file (`~/.aws/config`), a method of sourcing credentials from an external process. This can
+/// potentially be dangerous, so proceed with caution. Other credential providers should be
+/// preferred if at all possible. If using this option, you should make sure that the config file
+/// is as locked down as possible using security best practices for your operating system.
+///
+/// [credential_process]: https://docs.aws.amazon.com/cli/latest/topic/config-vars.html#sourcing-credentials-from-external-processes
+#[derive(Debug, Clone)]
+pub struct ChainProvider {
+    environment_provider: EnvironmentProvider,
+    instance_metadata_provider: InstanceMetadataProvider,
+    container_provider: ContainerProvider,
+    profile_provider: Option<ProfileProvider>,
+}
+
+impl ChainProvider {
+    /// Set the timeout on the provider to the specified duration.
+    #[allow(dead_code)]
+    pub fn set_timeout(&mut self, duration: Duration) {
+        self.instance_metadata_provider.set_timeout(duration);
+        self.container_provider.set_timeout(duration);
+    }
+}
+
+async fn chain_provider_credentials(
+    provider: ChainProvider,
+) -> Result<AwsCredentials, CredentialsError> {
+    if let Ok(creds) = provider.environment_provider.credentials().await {
+        return Ok(creds);
+    }
+    if let Some(ref profile_provider) = provider.profile_provider {
+        if let Ok(creds) = profile_provider.credentials().await {
+            return Ok(creds);
+        }
+    }
+    if let Ok(creds) = provider.container_provider.credentials().await {
+        return Ok(creds);
+    }
+    if let Ok(creds) = provider.instance_metadata_provider.credentials().await {
+        return Ok(creds);
+    }
+    if let Ok(creds) = WebIdentityProvider::from_k8s_env().credentials().await {
```

Comment:
I think we want to hold on to an instance of `WebIdentityProvider` in `ChainProvider`, so that the `WebIdentityProvider` can cache the OIDC token. That said all that amounts to is more calls to `os.Env` so maybe stateless is better!

Reply:
Good catch, I was wondering why what I wrote didn't match the existing pattern. Fixed.